File: backend/app/services/query_translator.py
from typing import List, Dict

class QueryTranslator:
    def to_openalex(self, query_blocks: List[Dict]) -> str:
        filters = []
        for block in query_blocks:
            field = block.get("field")
            value = block.get("value")
            type_ = block.get("type")
            
            if field == "title":
                filters.append(f"title.search:{value}")
            elif field == "year":
                if type_ == "greater_than":
                    filters.append(f"publication_year:>{value}")
                elif type_ == "less_than":
                    filters.append(f"publication_year:<{value}")
                else:
                    filters.append(f"publication_year:{value}")
        
        return ",".join(filters)

    def to_crossref(self, query_blocks: List[Dict]) -> Dict[str, str]:
        params = {}
        filters = []
        for block in query_blocks:
            field = block.get("field")
            value = block.get("value")
            type_ = block.get("type")
            
            if field == "title":
                params["query.title"] = value
            elif field == "year":
                if type_ == "equals":
                    filters.append(f"from-pub-date:{value}")
                    filters.append(f"until-pub-date:{value}")
                elif type_ == "greater_than":
                    filters.append(f"from-pub-date:{int(value) + 1}")
                elif type_ == "less_than":
                    filters.append(f"until-pub-date:{int(value) - 1}")
        
        if filters:
            params["filter"] = ",".join(filters)
            
        return params

File: frontend/electron/services/QueryTranslator.ts
import { QueryBlock } from './types';

export class QueryTranslator {
  toOpenAlex(queryBlocks: QueryBlock[]): string {
    const filters: string[] = [];
    for (const block of queryBlocks) {
      if (block.field === 'title') {
        filters.push(`title.search:${block.value}`);
      } else if (block.field === 'year') {
        if (block.type === 'greater_than') {
          filters.push(`publication_year:>${block.value}`);
        } else if (block.type === 'less_than') {
          filters.push(`publication_year:<${block.value}`);
        } else {
          filters.push(`publication_year:${block.value}`);
        }
      }
    }
    return filters.join(',');
  }

  toCrossref(queryBlocks: QueryBlock[]): Record<string, string> {
    const params: Record<string, string> = {};
    const filters: string[] = [];
    for (const block of queryBlocks) {
      if (block.field === 'title') {
        params['query.title'] = block.value;
      } else if (block.field === 'year') {
        if (block.type === 'equals') {
          filters.push(`from-pub-date:${block.value}`);
          filters.push(`until-pub-date:${block.value}`);
        } else if (block.type === 'greater_than') {
          filters.push(`from-pub-date:${parseInt(block.value) + 1}`);
        } else if (block.type === 'less_than') {
          filters.push(`until-pub-date:${parseInt(block.value) - 1}`);
        }
      }
    }
    if (filters.length > 0) {
      params['filter'] = filters.join(',');
    }
    return params;
  }
}
